reference-library/BeforeSwapDelta.md

```diff
@@ -0,0 +1,218 @@
+
+## Implementation Details
+
+The `BeforeSwapDelta` type and its associated functions use low-level assembly code for efficient bit manipulation and gas optimization:
+
+- The `toBeforeSwapDelta` function uses bitwise operations (`shl`, `or`, `and`, `sub`) to pack two `int128` values into a single `int256`.
+- The `getSpecifiedDelta` function uses the `sar` (shift arithmetic right) operation to extract the upper 128 bits.
+- The `getUnspecifiedDelta` function uses the `signextend` operation to extract and sign-extend the lower 128 bits.
+
+The `toBeforeSwapDelta` function combines the specified and unspecified deltas into a single `int256` value using bitwise operations. The `getSpecifiedDelta` and `getUnspecifiedDelta` functions extract the respective deltas using bit shifting and sign extension.
+
+By leveraging this compact representation and efficient arithmetic operations, Uniswap V4 can perform complex balance calculations and updates in a gas-optimized manner, reducing the overall cost of executing pool-related operations.
+
+## Comparison with BalanceDelta
+
+`BeforeSwapDelta` shares a similar structure with `BalanceDelta`, both packing two `int128` values into a single `int256`. However, there are key differences:
+
+- `BalanceDelta` represents amount0 and amount1.
+- `BeforeSwapDelta` represents specified and unspecified amounts, which may not directly correspond to token0 and token1, depending on the swap direction.
+
+## Best Practices
+
+When working with `BeforeSwapDelta`, consider the following best practices:
+
+- Always use the provided library functions (`getSpecifiedDelta` and `getUnspecifiedDelta`) to extract delta values.
+- Ensure that the signs of the delta values are correct from the hook's perspective.
+- Use `SafeCast` when converting between different integer types to prevent overflow/underflow errors.
+
+## Error Handling and Edge Cases
+
+- **Overflow/Underflow:** Ensure that the input int128 values do not exceed their range when packing into BeforeSwapDelta.
+- **Zero Values:** ZERO_DELTA represents no balance changes. Be cautious when interpreting zero values in specific contexts.
+- **Sign Mismatch:** Ensure that the signs of the delta values correctly represent the intended balance changes from the hook's perspective.
+
+## Example Usage in a Hook
+
+Here's a simple example of how `BeforeSwapDelta` might be used in a `beforeSwap` hook:
+
+```solidity
+function beforeSwap(address, PoolKey calldata, IPoolManager.SwapParams calldata params, bytes calldata)
+    external
+    override
+    returns (int256 amountIn, BeforeSwapDelta delta, uint24)
+{
+    int128 specifiedAmount = params.amountSpecified.toInt128();
+    int128 unspecifiedAmount = 0; // Calculated based on your custom logic
+
+    delta = toBeforeSwapDelta(specifiedAmount, unspecifiedAmount);
+    amountIn = params.amountSpecified;
+
+    return (amountIn, delta, 0);
+}
```

Comment:
can we explain what this is doing? its more helpful to show how to use it & explain what it does